torchao/prototype/low_bit_optim/cpu_offload.py

```diff
@@ -0,0 +1,101 @@
+from typing import Type
+
+import torch
+from torch.optim.optimizer import Optimizer
+
+
+class CPUOffloadOptimizer:
+    def __init__(self, params, optimizer_class: Type[Optimizer], *, offload_gradients: bool = False, **kwargs) -> None:
+        """Offload optimizer to CPU for single-GPU training. This will reduce GPU memory by the size of optimizer state.
+        Optimizer step will be done on CPU.
+
+        Args
+            params: a list of parameters or parameter groups.
+            optimizer_class: constructor of the base optimizer.
+            offload_gradients: free GPU gradients once they are moved to CPU. Not compatible with gradient accumulation.
+            kwargs: other keyword arguments to be passed to the base optimizer e.g. `lr`, `weight_decay`.
+        """
+        param_groups = list(params)
+        if len(param_groups) == 0:
+            raise ValueError("optimizer got an empty parameter list")
+        if not isinstance(param_groups[0], dict):
+            param_groups = [{"params": param_groups}]
+
+        self.param_cuda2cpu_map = dict()
+        self.optim_dict = dict()
+        self.stream = torch.cuda.Stream()
+
+        # the queue maintains the order which param we should do optim step on first.
+        self.queue = dict()
+
+        def backward_hook(p_cuda):
+            if p_cuda.grad is not None:
+                p_cpu = self.param_cuda2cpu_map[p_cuda]
+
+                # make sure backward for this param finishes
+                self.stream.wait_stream(torch.cuda.current_stream())
+                with torch.cuda.stream(self.stream):
+                    p_cpu.grad.copy_(p_cuda.grad, non_blocking=True)
+
+                # we rely on CPython implementation of dictionary, which preserves insertion order.
+                # if a param is added again (e.g. due to gradient accumulation), it is moved to the
+                # end of the queue by removing and inserting it again.
+                if p_cuda in self.queue:
+                    del self.queue[p_cuda]
+                self.queue[p_cuda] = self.stream.record_event()
+
+                # deallocate CUDA gradients once D2H transfer finishes.
+                if offload_gradients:
+                    p_cuda.grad.record_stream(self.stream)
+                    p_cuda.grad = None
+
+        for param_group in param_groups:
+            params = param_group.pop("params")
+
+            for p_cuda in params:
+                p_cpu = p_cuda.detach().cpu().pin_memory()
```

Comment:
nit: If you want this init to be slightly faster, you can probably pre-allocate the pinned memory and copy to it directly so that you do not have the intermediate copy to CPU paged memory.